File: FFCG.G4.GameOfLife/FFCG.G4.GameOfLife/GameOfLife.cs
using System;
using System.Collections.Generic;
using FFCG.G4.GameOfLife.Rules;

namespace FFCG.G4.GameOfLife
{
    public class GameOfLife
    {
        private readonly int _rowCount;
        private readonly int _columnCount;

        public GameOfLife(int rowCount, int columnCount)
        {
            _rowCount = rowCount;
            _columnCount = columnCount;

            CreateCells();
        }

        public Cell[,] Cells { get; private set; }

        private void CreateCells()
        {
            Cells = new Cell[_rowCount, _columnCount];

            var lifeRules = new List<ILifeRule>()
            {
                new LiveCellWithMoreThanThreeNeighborsShouldDieRule(),
                new LiveCellWithFewerThanTwoNeighborsRule(),
                new LiveCellWithTwoOrThreeNeighborsShouldLiveRule(),
                new DeadCellWithThreeNeighborsShouldLiveRule()
            };

            for (var x = 0; x < _rowCount; x++)
                for (var y = 0; y < _columnCount; y++)
                    Cells[x, y] = new Cell(lifeRules);
        }

        public void SetLivingCell(int row, int column)
        {
            Cells[row, column].IsAlive = true;
        }

        public void ComputeNextGeneration()
        {
            var nextGeneration = new Cell[_rowCount, _columnCount];

            for (var x = 0; x < _rowCount; x++)
                for (var y = 0; y < _columnCount; y++)
                    nextGeneration[x, y] = Cells[x, y].Split(CountLivingNeighbors(x, y));

            Cells = nextGeneration;
        }

        public int CountLivingNeighbors(int row, int column)
        {
            var surroundingCellPositions = new List<Tuple<int, int>>()
            {
                new Tuple<int, int>(row-1, column-1),
                new Tuple<int, int>(row-1, column),
                new Tuple<int, int>(row-1, column+1),
                new Tuple<int, int>(row, column+1),
                new Tuple<int, int>(row+1, column+1),
                new Tuple<int, int>(row+1, column),
                new Tuple<int, int>(row+1, column-1),
                new Tuple<int, int>(row, column-1)
            };

            var livingNeighbors = 0;

            foreach (var position in surroundingCellPositions)
                if (IsInTheGrid(position.Item1, position.Item2) && Cells[position.Item1, position.Item2].IsAlive)
                    livingNeighbors++;

            return livingNeighbors;
        }

        private bool IsInTheGrid(int row, int col)
        {
            return row >= 0 && col >= 0 && row < _rowCount && col < _columnCount;
        }
    }
}

File: FFCG.G7.GameOfLife/src/GameOfLife.ts
import { Cell } from "./cell";
import { ILifeRule } from "./rules/ILifeRule";
import { DeadCellWithThreeNeighbors } from "./rules/DeadCellWithThreeNeighbors";
import { LiveCellWithFewerThanTwoNeighborsRule } from "./rules/LiveCellWithFewerThanTwoNeighborsRule";
import { LiveCellWithMoreThanThreeNeighborsRule } from "./rules/LiveCellWithMoreThanThreeNeighborsRule";
import { LiveCellWithTwoOrThreeNeighborsRule } from "./rules/LiveCellWithTwoOrThreeNeighborsRule";

export class GameOfLife {
    rows: number
    columns: number

    cells: Array<Array<Cell>> = []
    lifeRules: Array<ILifeRule>

    constructor(rows: number, columns: number) {
        this.rows = rows
        this.columns = columns
        this.createCells()
    }

    private createCells(): void {
        this.lifeRules = [
            new DeadCellWithThreeNeighbors(),
            new LiveCellWithFewerThanTwoNeighborsRule(),
            new LiveCellWithMoreThanThreeNeighborsRule(),
            new LiveCellWithTwoOrThreeNeighborsRule()
        ]

        for (var x = 0; x < this.rows; x++) {
            this.cells[x] = []
            
            for (var y = 0; y < this.columns; y++)
                this.cells[x][y] = new Cell(this.lifeRules)
        }
    }

    setLivingCell = (x: number, y: number) => this.cells[x][y].isAlive = true

    countLivingNeighbors(x: number, y: number) {
        const surroundingCellPositions = [
            [x - 1, y - 1],
            [x - 1, y],
            [x - 1, y + 1],
            [x, y - 1],
            [x, y + 1],
            [x + 1, y - 1],
            [x + 1, y],
            [x + 1, y + 1]
        ]

        return surroundingCellPositions.filter(p => this.isWithinGrid(p[0], p[1]) && this.cells[p[0]][p[1]].isAlive).length
    }

    private isWithinGrid = (row: number, column: number) => row >= 0 && column >= 0 && row < this.rows && column < this.columns

    computeNextGeneration() {
        const nextGeneration = []

        for (var x = 0; x < this.rows; x++) {
            nextGeneration[x] = []
            
            for (var y = 0; y < this.columns; y++)
            nextGeneration[x][y] = this.cells[x][y].next(this.countLivingNeighbors(x, y))
        }

        this.cells = nextGeneration
    }
}